app/forms/csv_upload_form.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+class CsvUploadForm
+  include ActiveModel::Validations
+
+  attr_reader :file
+
+  validates_with CsvValidator
+
+  def initialize(current_user)
+    @current_user = current_user
+  end
+
+  def save(file)
+    @file = file
+    return false unless file && valid?
+
+    # rubocop:disable Rails/SkipsModelValidations
+    Keyword.insert_all(keyword_hash)
+    # rubocop:enable Rails/SkipsModelValidations
+
+    errors.empty?
+  end
+
+  private
+
+  def keyword_hash
+    CSV.read(file).filter_map do |row|
+      name = row.join(',')
+      name.blank? ? false : { name: name, user_id: @current_user.id }
+    end
```

Comment:
To be honest, I did not understand the code until I debug it and run a test to see what `row` contained 🥲 

```suggestion
    CSV.read(file).filter_map do |row_columns|
      name = row_columns.to_csv.chomp
      name.blank? ? false : { name: name, user_id: @current_user.id }
    end
```


Also, as `insert_all` is skipping validations, should we build and validate a keyword object at this stage? 🤔 

That would give us:

```
  def keywords
    CSV.read(file).filter_map do |row_columns|
      keyword = Keyword.new(name: row_columns.to_csv.chomp, user_id: @current_user.id)

      keyword if keyword.valid?
    end
  end
```

and so the save method can be simplified to just save keywords (a little bit less performant*** but will still run a single DB query):

> _*** due the performance reduction is actually due to the instantiation and validation made inside the `keywords` method_

```
  def save(file)
    @file = file
    return false unless file && valid?

    ActiveRecord::Base.transaction { keywords.each(&:save) }

    errors.empty?
  end
```

The tests are still passing with this code suggestion so I think it's OK.

Reply:
I compared `single save` and `import_all`, the performance seems similar with low entries count <1000. 👍